### PR TITLE
Add kpod load

### PR DESCRIPTION
**Pull request by umohnani8**
Add kpod load. Add fixes to kpod pull. Add changes to containers/image/docker to get manifest for image:tag.

### CODE REVIEW

**Comment by umohnani8** [comment]
Created a pull request for the change to containers/image https://github.com/containers/image/pull/309

**Comment by umohnani8** [comment]
Need kpod save to write tests. Will do that once I finish kpod save.

File: cmd/kpod/load.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+type loadOptions struct {
+	oci   bool
+	input string
+	quiet bool
+}
+
+var (
+	loadFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "oci",
```
Comment (rhatdan) [inline cmd/kpod/load.go]: Any way we could determine this from the image, rather then making the user specify which type he wants?

File: cmd/kpod/load.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			Usage: "Load image into oci",
+		},
+		cli.StringFlag{
+			Name:  "input, i",
+			Usage: "Read from archive file, instead of STDIN",
```
Comment (rhatdan) [inline cmd/kpod/load.go]: default is STDIN

File: cmd/kpod/pull.go [hunk 3]
```diff
@@ -71,42 +75,54 @@ func pullCmd(c *cli.Context) error {
...
+		if err != nil {
+			return errors.Errorf("error retrieving manifest.json: %v", err)
+		}
+		// to pull all the images stored in one tar file
+		for i := range manifest {
+			images = append(images, manifest[i].RepoTags[0])
+		}
+	} else if splitArr[0] == "oci" {
+		// needs to be implemented
+		fmt.Println("oci")
```
Comment (rhatdan) [inline cmd/kpod/pull.go]: Debug?

File: docs/kpod-load.1.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+**--oci**
+Load image into oci
+
+## GLOBAL OPTIONS
+
+**--help, -h**
+  Print usage statement
+
+## SEE ALSO
+kpod(1), crio(8), crio.conf(5)
```
Comment (rhatdan) [inline docs/kpod-load.1.md]: kpod-save(1)
I know it does not exist yet, but why wait

File: docs/kpod-load.1.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+**kpod [GLOBAL OPTIONS]**
+
+**kpod load [GLOBAL OPTIONS]**
+
+**kpod load [OPTIONS] NAME[:TAG|@DIGEST] [GLOBAL OPTIONS]**
+
+##OPTIONS
+
+**--input, -i**
+Read from archive file, instead of STDIN
```
Comment (rhatdan) [inline docs/kpod-load.1.md]: default is STDIN

File: docs/kpod-load.1.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+**--oci**
+Load image into oci
+
+## GLOBAL OPTIONS
+
+**--help, -h**
+  Print usage statement
+
+## SEE ALSO
```
Comment (rhatdan) [inline docs/kpod-load.1.md]:
```
## Examples
# kpod load -i fedora.tar

# cat /tmp/RHEL7.gzip | kpod load -q
```

**Comment by rhatdan** [comment]
Update README.md

File: cmd/kpod/load.go (hunk 1 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: I thought you wanted the oci flag so it loads from the oci archive when it is set. There is a typo in the Usage, should be "Load image from oci archive." I will fix that.

Comment (rhatdan) [inline cmd/kpod/load.go]: I think saving is fine, not sure we need one for load.  We could attempt to load it as an oci and then as an docker-archive.

Comment (umohnani8) [inline cmd/kpod/load.go]: What if the user has saved an image with the same name to docker-archive and oci, but wants to load the docker-archive one?

Comment (rhatdan) [inline cmd/kpod/load.go]: The user is handing the tool the archive, there is no choice.  
Think of it this way, this is like doing an editor

office resume.doc
office resume.ods

Would you make the user specify --doc, in order to read the .doc file or would the tool be smart enough to figure out that this is a .doc

File: cmd/kpod/load.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+		return errors.New("too many arguments. Requires exactly 1")
+	}
+
+	oci := c.Bool("oci")
+	input := c.String("input")
+	quiet := c.Bool("quiet")
+
+	opts := loadOptions{
+		oci:   oci,
+		input: input,
```
Comment (mrunalp) [inline cmd/kpod/load.go]: We should check the type of input and process accordingly. Best would be to pass a reader to the library:
```
diff --git a/cmd/kpod/load.go b/cmd/kpod/load.go
index ae346bc8..97783682 100644
--- a/cmd/kpod/load.go
+++ b/cmd/kpod/load.go
@@ -1,6 +1,8 @@
 package main

 import (
+       "fmt"
+       "os"
        "github.com/containers/storage"
        "github.com/pkg/errors"
        "github.com/urfave/cli"
@@ -59,6 +61,19 @@ func loadCmd(c *cli.Context) error {
        input := c.String("input")
        quiet := c.Bool("quiet")

+       if input == "/dev/stdin" {
+               fi, err := os.Stdin.Stat()
+               if err != nil {
+                       return err
+               }
+               if fi.Mode().IsRegular() {
+                       fmt.Println("Loading from regular file")
+               } else {
+                       fmt.Println("Loading from a pipe")
+               }
+       }
+
+
```

Comment (rhatdan) [inline cmd/kpod/load.go]: @umohnani8 has a version of this that creates a temporary file when this stdin is a pipe, and then hands that to the function call. Not sure if there is an interface that takes a reader rather then just a file.

File: cmd/kpod/load.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			}
+			defer inFile.Close()
+
+			_, err = io.Copy(outFile, inFile)
+			if err != nil {
+				return errors.Errorf("error copying file %v", err)
+			}
+
+			input = outFile.Name()
+			defer os.Remove(outFile.Name())
```
Comment (mrunalp) [inline cmd/kpod/load.go]: This could be moved further up after the close.

File: cmd/kpod/load.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	input := c.String("input")
+	quiet := c.Bool("quiet")
+
+	if input == "/dev/stdin" {
+		fi, err := os.Stdin.Stat()
+		if err != nil {
+			return err
+		}
+		if !fi.Mode().IsRegular() {
+			outFile, err := ioutil.TempFile("/tmp", "kpod")
```
Comment (mrunalp) [inline cmd/kpod/load.go]: Use some other directory as /tmp could use up RAM.

Comment (rhatdan) [inline cmd/kpod/load.go]: I think we should just switch to /var/tmp for this. /var/tmp/ is used for temporary files which can be large.

File: cmd/kpod/load.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+			Name:  "input, i",
+			Usage: "Read from archive file, default is STDIN",
+			Value: "/dev/stdin",
+		},
+		cli.BoolFlag{
+			Name:  "quiet, q",
+			Usage: "Suppress the output",
+		},
+	}
+	loadDescription = "Loads the image from docker-archive by default or from oci if the oci flag is set."
```
Comment (rhatdan) [inline cmd/kpod/load.go]: There is no OCI flag.  Remove this qualifier.  I think for now we should remove oci from this, until we have an oci-archive.  It seems that oci transport creates a directory structure not a tar ball.

File: cmd/kpod/load.go (hunk 9 above)
Comment (rhatdan) [inline cmd/kpod/load.go]: /var/tmp

File: cmd/kpod/load.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+// using the pullImage function
+func loadImage(store storage.Store, opts loadOptions) error {
+	var (
+		src string
+		err error
+	)
+
+	systemContext := getSystemContext("")
+
+	src = dockerArchive + opts.input
```
Comment (rhatdan) [inline cmd/kpod/load.go]: Nit
src := dockerArchive + opts.input

File: cmd/kpod/load.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+	if err = pullImage(store, src, false, opts.quiet, systemContext); err != nil {
```
Comment (rhatdan) [inline cmd/kpod/load.go]: err := ...

File: cmd/kpod/load.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+		src = oci + opts.input
+		if err = pullImage(store, src, false, opts.quiet, systemContext); err != nil {
+			return err
+		}
+	}
+	return err
```
Comment (rhatdan) [inline cmd/kpod/load.go]: Return nil

File: cmd/kpod/pull.go [hunk 14]
```diff
@@ -118,8 +129,16 @@ func pullImage(store storage.Store, imgName string, allTags bool, sc *types.Syst
...
+	for _, image := range images {
+		destRef, err := is.Transport.ParseStoreReference(store, image)
+		if err != nil {
+			return errors.Errorf("error parsing dest reference name: %v", err)
+		}
+		if err = cp.Image(policyContext, destRef, srcRef, getCopyOptions(output)); err != nil {
+			return errors.Errorf("error loading image %q: %v", image, err)
+		}
+	}
+	return err
```
Comment (rhatdan) [inline cmd/kpod/pull.go]: return nil

File: completions/bash/kpod [hunk 15]
```diff
@@ -93,6 +93,17 @@ _kpod_pull() {
     _complete_ "$options_with_args" "$boolean_options"
 }
 
+_kpod_load() {
+    local options_with_args="
+    "
+    local boolean_options="
+    --input
+    --quiet -q
+    --oci
```
Comment (rhatdan) [inline completions/bash/kpod]: No --oci

File: cmd/kpod/load.go (hunk 8 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: Moved up.

File: cmd/kpod/load.go (hunk 9 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: using /var/tmp

File: cmd/kpod/load.go (hunk 10 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: Removed everything to do with OCI

File: cmd/kpod/load.go (hunk 11 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: done

File: cmd/kpod/load.go (hunk 12 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: done

File: cmd/kpod/load.go (hunk 13 above)
Comment (umohnani8) [inline cmd/kpod/load.go]: done

File: cmd/kpod/pull.go (hunk 14 above)
Comment (umohnani8) [inline cmd/kpod/pull.go]: done

File: completions/bash/kpod (hunk 15 above)
Comment (umohnani8) [inline completions/bash/kpod]: done

**Comment by rhatdan** [comment]
https://github.com/containers/image/pull/309 has been merged upstream, now we need to vendor this in.

**Comment by mrunalp** [comment]
Needs rebase

File: vendor/github.com/containers/image/docker/tarfile/src.go [hunk 16]
```diff
@@ -146,22 +146,22 @@ func (s *Source) ensureCachedDataIsPresent() error {
 	}
```
Comment (mrunalp) [inline vendor/github.com/containers/image/docker/tarfile/src.go]: I don't see a vendor.conf update corresponding to this.

Comment (umohnani8) [inline vendor/github.com/containers/image/docker/tarfile/src.go]: @rhatdan has already done it https://github.com/kubernetes-incubator/cri-o/pull/679

Comment (mrunalp) [inline vendor/github.com/containers/image/docker/tarfile/src.go]: okay, we should add the top comment here saying don't merge till #679 is in.

Comment (rhatdan) [inline vendor/github.com/containers/image/docker/tarfile/src.go]: @umohnani8 add [WIP] or [DONTMERGE] to heading until this gets vendored in.

**Comment by rhatdan** [comment]
Lots of lint failures.

**Comment by rhatdan** [comment]
LGTM

**Comment by umohnani8** [comment]
rebased. @mrunalp PTAL.

**Comment by mrunalp** [comment]
The commits below this are more than 10 days old. Could you rebase to latest master?

**Comment by umohnani8** [comment]
Rebased

**Comment by mrunalp** [comment]
Cool, we need to wait for the tests again.

**Comment by rhatdan** [comment]
@mrunalp tests passed, this is ready to merge.

**Pull request merged**